### ISSUE TITLE
Eager-loaded nil associations result in N+1 query problem on Rails 4.1 & 4.2 when using `references`

### ISSUE THREAD

**Issue by codeodor** [rails/rails#18312]
Suppose you have

```ruby
class Season < ActiveRecord::Base
  belongs_to :division
  has_one :champion
end

class Division < ActiveRecord::Base
end

class Champion < ActiveRecord::Base
end
```

Then we have this code that uses it:

```ruby
seasons = Season.includes(:division, :champion).references(:champion)
seasons.each{|s| puts s.champion.inspect }
```

This will run a query for each `champion` if `champion` is `nil` starting in Rails 4.1.

If you remove the call to `references` however, it behaves as expected.

**Comment by codeodor** [comment]
I can fix the behavior by messing around here: https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/associations/join_dependency.rb#L252

Basically calling `ar_parent.association(node.reflection.name).target = nil` before `next` so that [`association_instance_set`](https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/associations.rb#L176) is called and a key for the `@association_cache` gets set, which will make it not try to reload.

However, that causes many more problems throughout the tests, so it's not the correct fix, just a spot to start looking.

**Comment by codeodor** [comment]
In the case with `references` it constructs the association in `join_dependency.rb` while without `references` it constructs it in `preloader.rb`.

**Comment by codeodor** [comment]
If I change what I said above to not set the target to nil (i.e. just do `ar_parent.association(node.reflection.name)`, it seems to fix the issue a little better, but there are extra items returned for some of the eager loading test cases which are collections.

**Comment by codeodor** [comment]
Fixed with #18316
